dojo/tools/api_msdefender/parser.py

```diff
@@ -0,0 +1,106 @@
+import json
+from dojo.models import Finding, Endpoint
+from .importer import MSDefenderApiImporter
+
+
+class ApiMSDefenderParser(object):
+    """
+    Import from MSDefender API /findings
+    """
+
+    def get_scan_types(self):
+        return ["MSDefender API"]
+
+    def get_label_for_scan_types(self, scan_type):
+        return "MSDefender API"
+
+    def get_description_for_scan_types(self, scan_type):
+        return ("MSDefender findings can be directly imported using the REST API")
+
+    def requires_file(self, scan_type):
+        return False
+
+    def requires_tool_type(self, scan_type):
+        return "MSDefender API"
+
+    def api_scan_configuration_hint(self):
+        return (
+            "Do not set anything here."
+        )
+
+    def get_findings(self, file, test):
+        if file:
+            data = []
+            vulnerabilityfile = json.load(file[0])
+            vulnerabilitydata = vulnerabilityfile['value']
+            data.append(vulnerabilitydata)
+            machinefile = json.load(file[1])
+            machinedata = machinefile['value']
+            data.append(machinedata)
+        else:
+            data = MSDefenderApiImporter().get_findings(test)
+        return self.process_vulnerabilities(test, data)
+
+    def machinelisttranslator(self, machinelist):
+        self.dictformachinelist = {}
+        i = 0
+        for machines in machinelist:
+            self.dictformachinelist[machines['id']] = i
+            i += 1
+
+    def returnlistfromtranslator(self, machineid, machinelist):
+        return machinelist[self.dictformachinelist[machineid]]
+
+    def process_vulnerabilities(self, file, test):
+        self.machinelisttranslator(machinelist=test[1])
+        findings = []
+        for vulnerability in test[0]:
+            machine = self.returnlistfromtranslator(machineid=vulnerability['machineId'], machinelist=test[1])
```

Comment:
In my opinion, you can drop `machinelisttranslator`, `returnlistfromtranslator` and `dictformachinelist`.

Simple lookup `machine = list(filter(lambda m: m['id'] == vulnerability['machineId'], test[1]))[0]` will do the job.

Reply:
privatesdefectdojo-uwsgi-1         |   File "/app/dojo/importers/importer/importer.py", line 315, in import_scan
privatesdefectdojo-uwsgi-1         |     parsed_findings = parser.get_findings(scan, test)
privatesdefectdojo-uwsgi-1         |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
privatesdefectdojo-uwsgi-1         |   File "/app/dojo/tools/api_msdefender/parser.py", line 42, in get_findings
privatesdefectdojo-uwsgi-1         |     return self.process_vulnerabilities(data)
privatesdefectdojo-uwsgi-1         |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
privatesdefectdojo-uwsgi-1         |   File "/app/dojo/tools/api_msdefender/parser.py", line 47, in process_vulnerabilities
privatesdefectdojo-uwsgi-1         |     machine = list(filter(lambda m: m['id'] == vulnerability['machineId'], data[1]))[0]
privatesdefectdojo-uwsgi-1         |               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^
privatesdefectdojo-uwsgi-1         | IndexError: list index out of range